### PR TITLE
Fix(47969): String.prototypr.replace docs fix

### PR DESCRIPTION
**Pull request by navya9singh**
Fixes https://github.com/microsoft/TypeScript/issues/47969

### CODE REVIEW

**Comment by amcasey** [comment]
@sandersn Shouldn't this be `Author: Team`?

File: src/lib/es2015.symbol.wellknown.d.ts [hunk 1]
```diff
@@ -219,9 +219,9 @@ interface String {
     match(matcher: { [Symbol.match](string: string): RegExpMatchArray | null; }): RegExpMatchArray | null;
 
     /**
-     * Replaces first match with string or all matches with RegExp.
-     * @param searchValue A string or RegExp search value.
-     * @param replaceValue A string containing the text to replace for match.
+     * Replaces one or more occurrences of substrings that match the method provided by `searchValue`.
+     * @param searchValue An object that supports searching for and replacing matches within a string.
+     * @param replacer The replacement text.
```
Comment (amcasey) [inline src/lib/es2015.symbol.wellknown.d.ts:224]: It looks like the parameter might be named `replaceValue`?

File: src/lib/es2015.symbol.wellknown.d.ts [hunk 2]
```diff
@@ -219,9 +219,9 @@ interface String {
     match(matcher: { [Symbol.match](string: string): RegExpMatchArray | null; }): RegExpMatchArray | null;
 
     /**
-     * Replaces first match with string or all matches with RegExp.
-     * @param searchValue A string or RegExp search value.
-     * @param replaceValue A string containing the text to replace for match.
+     * Replaces one or more occurrences of substrings that match the method provided by `searchValue`.
```
Comment (amcasey) [inline src/lib/es2015.symbol.wellknown.d.ts:222]: I'm surprised that it will replace multiple instances, but not based on any actual experience.  You confirmed somehow?

File: src/lib/es5.d.ts [hunk 3]
```diff
@@ -430,7 +430,7 @@ interface String {
     /**
      * Replaces text in a string, using a regular expression or search string.
      * @param searchValue A string to search for.
```
Comment (amcasey) [inline src/lib/es5.d.ts:432]: Should this say something about "or regular expression"?

File: src/lib/es5.d.ts [hunk 4]
```diff
@@ -430,7 +430,7 @@ interface String {
     /**
      * Replaces text in a string, using a regular expression or search string.
      * @param searchValue A string to search for.
-     * @param replaceValue A string containing the text to replace for every successful match of searchValue in this string.
+     * @param replaceValue A string containing the text to replace. When the searchvalue is a string, only the first match is replaced. If the searchValue is a Regexp, all matches are replaced if the g flag is set. Otherwise only the first one is.
```
Comment (amcasey) [inline src/lib/es5.d.ts:433]: I think there might be syntax for referring to another parameter (i.e. in a way that will work with find-all-refs or rename) - @sandersn  would know.

**Review by amcasey — CHANGES_REQUESTED** [review]

File: src/lib/es5.d.ts (hunk 4 above)
Comment (Josh-Cena) [inline src/lib/es5.d.ts:433]: There's an "interesting" aspect where `"aaa-a".replace(/a/gy, "x")` is `xxx-a`. I don't know how detailed we want it to be, but "all matches are replaced if the g flag is set" is technically not correct.

File: src/lib/es2015.symbol.wellknown.d.ts (hunk 2 above)
Comment (Josh-Cena) [reply src/lib/es2015.symbol.wellknown.d.ts:222]: This is the doc for `@@replace` so it can have any custom behavior—only when `searchValue` is a string will it only replace once.

File: src/lib/es5.d.ts (hunk 4 above)
Comment (sandersn) [reply src/lib/es5.d.ts:433]: The syntax is `{@link searchValue}`, although maybe `{@linkcode searchValue}` would be better since it usually produces the `monospace` font used for code.

File: src/lib/es2015.symbol.wellknown.d.ts (hunk 2 above)
Comment (sandersn) [reply src/lib/es2015.symbol.wellknown.d.ts:222]: The custom behaviour is the point, so I think it'd be better to talk about that. Maybe:

```suggestion
     * Passes a string and {@linkcode replaceValue} to the [Symbol.replace] method on {@linkCode searchValue}. This method is expected to implement its own replacement algorithm.
```

I'm not sure whether

1. `[Symbol.replace]` is enough explanation of how to construct a good searchValue.
2. It's actually true that the method is expected to implement its own replacement algorithm.

**Review by sandersn — CHANGES_REQUESTED** [review]

File: src/lib/es2015.symbol.wellknown.d.ts (hunk 2 above)
Comment (Josh-Cena) [reply src/lib/es2015.symbol.wellknown.d.ts:222]: Is it possible that we add another `replacementValue: RegExp` overload before this? FWIW, replacing with regexes remains what people write 99% of the time and it makes sense to have some doc more specific about that.

> It's actually true that the method is expected to implement its own replacement algorithm.

I believe `@@replace` was implemented purely to make string methods decoupled from regex semantics. Nevertheless, yes, it can have any custom behavior.

**Comment by orta** [comment]
>  Shouldn't this be Author: Team?

edit: https://github.com/microsoft/TypeScript/blob/main/.github/pr_owners.txt to update this (comes from [here](https://github.com/microsoft/TypeScript-repos-automation/blob/master/src/pr_meta/isMemberOfTSTeam.ts))

**Comment by amcasey** [comment]
> > Shouldn't this be Author: Team?
> 
> edit: https://github.com/microsoft/TypeScript/blob/main/.github/pr_owners.txt to update this (comes from [here](https://github.com/microsoft/TypeScript-repos-automation/blob/master/src/pr_meta/isMemberOfTSTeam.ts))

Done.  Thanks, @orta!

File: src/lib/es5.d.ts (hunk 4 above)
Comment (sandersn) [reply src/lib/es5.d.ts]: MDN's documentation for the sticky flag says that it ignores the global flag. If that's true, then I think it's fine to document it as-is, since you'll presumably find out about `y` long after `g` and hopefully read about the override.

However, is that consistent with your example? If there's some kind of interaction, it *might* be worth documenting.

**Comment by sandersn** [comment]
Looks pretty good but I'd like to address @Josh-Cena 's point about the sticky flag before merging.

File: src/lib/es5.d.ts (hunk 4 above)
Comment (Josh-Cena) [reply src/lib/es5.d.ts]: There's an open issue for it: https://github.com/mdn/content/issues/1454 I just haven't gotten around to fix it.

File: src/lib/es2015.symbol.wellknown.d.ts [hunk 5]
```diff
@@ -219,9 +219,9 @@ interface String {
     match(matcher: { [Symbol.match](string: string): RegExpMatchArray | null; }): RegExpMatchArray | null;
 
     /**
-     * Replaces first match with string or all matches with RegExp.
-     * @param searchValue A string or RegExp search value.
-     * @param replaceValue A string containing the text to replace for match.
+     * Passes a string and {@linkcode replaceValue} to the [Symbol.replace] method on {@linkcode searchValue}. This method is expected to implement its own replacement algorithm.
```
Comment (Josh-Cena) [inline src/lib/es2015.symbol.wellknown.d.ts:222]:
```suggestion
     * Passes a string and {@linkcode replaceValue} to the `[Symbol.replace]` method on {@linkcode searchValue}. This method is expected to implement its own replacement algorithm.
```

(Suggestion, dunno if the stdlib has its own style rules about JSDoc. Also `@@replace` is technically the name for the method—it's only accessed via `Symbol.replace`)

Comment (sandersn) [reply src/lib/es2015.symbol.wellknown.d.ts:222]: I like the formatting improvement, but I think `@@replace` is only useful for people who have read the spec.

Comment (Josh-Cena) [reply src/lib/es2015.symbol.wellknown.d.ts:222]: Yeah, I'm fine with `[Symbol.replace]`. FWIW, on MDN we consistently use `RegExp.prototype[@@replace]()` so I hope it's not _that_ obscure, but I can understand.

File: src/lib/es5.d.ts [hunk 6]
```diff
@@ -429,8 +429,8 @@ interface String {
 
     /**
      * Replaces text in a string, using a regular expression or search string.
-     * @param searchValue A string to search for.
-     * @param replaceValue A string containing the text to replace for every successful match of searchValue in this string.
+     * @param searchValue A string or regular expression to search for.
+     * @param replaceValue A string containing the text to replace. When the {@linkcode searchValue} is a string, only the first match is replaced. If the {@linkcode searchValue} is a RegExp, all matches are replaced if the g flag is set. Otherwise only the first one is.
```
Comment (sandersn) [inline src/lib/es5.d.ts:433]: I think I would reword this to highlight the interesting global regex case first, and try to cover strings and sticky regexes later.

```suggestion
     * @param replaceValue A string containing the text to replace. When the {@linkcode searchValue} is a RegExp, all matches are replaced if the g flag is set (taking into account the y flag if present). Otherwise {@linkcode searchValue} only the first match is replaced.
```

File: src/lib/es5.d.ts [hunk 7]
```diff
@@ -429,8 +429,8 @@ interface String {
 
     /**
      * Replaces text in a string, using a regular expression or search string.
-     * @param searchValue A string to search for.
-     * @param replaceValue A string containing the text to replace for every successful match of searchValue in this string.
+     * @param searchValue A string or regular expression to search for.
+     * @param replaceValue A string containing the text to replace. When the {@linkcode searchValue} is a RegExp, all matches are replaced if the g flag is set (taking into account the y flag if present). Otherwise {@linkcode searchValue} only the first match is replaced.
```
Comment (Josh-Cena) [inline src/lib/es5.d.ts:433]:
```suggestion
     * @param replaceValue A string containing the text to replace. When the {@linkcode searchValue} is a `RegExp`, all matches are replaced if the `g` flag is set (or only those matches at the beginning, if the `y` flag if present). Otherwise, only the first match of {@linkcode searchValue} is replaced.
```

Nit

Comment (sandersn) [reply src/lib/es5.d.ts]: I like this wording better, but note that the bit before the close paren should swap if->is: `flag is present` not `flag if present`.

**Review by sandersn — APPROVED** [review]
Once @Josh-Cena's latest suggestion is adopted (along with my if -> is fix)

File: src/lib/es5.d.ts (hunk 7 above)
Comment (Josh-Cena) [reply src/lib/es5.d.ts]: Ah yes—let's say "is also".

**Comment by graphemecluster** [comment]
@sandersn I can hardly understand that how come the documentation of only 2 out of 4 overloads are updated.
And `replaceAll` is not touched at all.
Anyway I have updated them in my PR (#50452).
